ext/node/zlib/mod.rs

```diff
@@ -0,0 +1,474 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::bad_resource_id;
+use deno_core::error::type_error;
+use deno_core::error::AnyError;
+use deno_core::op;
+use deno_core::OpState;
+use deno_core::ZeroCopyBuf;
+use libz_sys::*;
+use std::borrow::Cow;
+use std::cell::RefCell;
+use std::future::Future;
+use std::rc::Rc;
+
+mod alloc;
+
+#[inline]
+fn check(condition: bool, msg: &str) -> Result<(), AnyError> {
+  if condition {
+    Ok(())
+  } else {
+    Err(type_error(msg.to_string()))
+  }
+}
+
+struct ZlibInner {
+  dictionary: Option<Vec<u8>>,
+  err: i32,
+  flush: i32,
+  init_done: bool,
+  level: i32,
+  mem_level: i32,
+  mode: i32,
+  strategy: i32,
+  window_bits: i32,
+  write_in_progress: bool,
+  pending_close: bool,
+  gzib_id_bytes_read: u32,
+  strm: z_stream,
+}
+
+const GZIP_HEADER_ID1: u8 = 0x1f;
+const GZIP_HEADER_ID2: u8 = 0x8b;
+
+impl ZlibInner {
+  fn start_write(
+    &mut self,
+    input: &[u8],
+    in_off: u32,
+    in_len: u32,
+    out: &mut [u8],
+    out_off: u32,
+    out_len: u32,
+    flush: i32,
+  ) -> Result<(), AnyError> {
+    check(self.init_done, "write before init")?;
+    check(self.mode != NONE, "already finialized")?;
+    check(!self.write_in_progress, "write already in progress")?;
+    check(!self.pending_close, "close already in progress")?;
+
+    self.write_in_progress = true;
+
+    if flush != Z_NO_FLUSH
+      && flush != Z_PARTIAL_FLUSH
+      && flush != Z_SYNC_FLUSH
+      && flush != Z_FULL_FLUSH
+      && flush != Z_FINISH
+      && flush != Z_BLOCK
+    {
+      return Err(type_error("Bad argument"));
+    }
+
+    self.strm.avail_in = in_len;
+    // TODO(@littledivy): Crap! Guard against overflow!!
+    self.strm.next_in = unsafe { input.as_ptr().offset(in_off as isize) } as _;
+    self.strm.avail_out = out_len;
+    self.strm.next_out =
+      unsafe { out.as_mut_ptr().offset(out_off as isize) } as _;
+
+    self.flush = flush;
+    Ok(())
+  }
+
+  fn do_write(&mut self, flush: i32) -> Result<(), AnyError> {
+    match self.mode {
+      DEFLATE | GZIP | DEFLATERAW => {
+        self.err = unsafe { libz_sys::deflate(&mut self.strm, flush) };
```

Comment:
Could you move the `unsafe` `libz_sys::*` calls into functions in one spot of the file for easier inspection? I think it's a bit easier to inspect all unsafe code if it's grouped together.

Reply:
Moved to all libz unsafe calls to `stream.rs`